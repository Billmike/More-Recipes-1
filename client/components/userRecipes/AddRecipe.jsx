import React from 'react';
import PropTypes from 'prop-types';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import isEmpty from 'lodash/isEmpty';
import Dropzone from 'react-dropzone';

import Spinner from '../common/Spinner';
import ErrorMessage from '../common/ErrorMessage';

const propTypes = {
  category: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  handleSelect: PropTypes.func.isRequired,
  handleAddRecipe: PropTypes.func.isRequired,
  handleDrop: PropTypes.func.isRequired,
  imagePreview: PropTypes.string.isRequired,
  userRecipes: PropTypes.shape({
    isLoading: PropTypes.bool,
    imageUploading: PropTypes.bool.isRequired,
    error: PropTypes.shape()
  }).isRequired,
};

/**
 * Add recipe component
 *
 * @param {any} props - The props passed
 *
 * @returns { object } React component
 */
const AddRecipe = (props) => {
  const { category, open, handleChange, handleSelect, handleDrop, handleClose,
    handleAddRecipe, imagePreview, userRecipes } = props;
  const { isLoading, imageUploading, error } = userRecipes;
  const actions = [
    <FlatButton
      label="Cancel"
      secondary
      onClick={handleClose}
    />
  ];

  return (
    <div>
      <Dialog
        title="Add Recipe"
        actions={actions}
        modal
        open={open}
        autoScrollBodyContent
      >
        <div id="add-recipe" className="row">
          <form className="col s12" onSubmit={handleAddRecipe} >
            <div className="row">
              <div className="input-field col s12">
                <input
                  name="title"
                  type="text"
                  className="validate"
                  maxLength="255"
                  onChange={handleChange}
                  required
                />
                <label htmlFor="recipe_name">Title</label>
              </div>
            </div>
            <div className="row">
              <select
                className="browser-default"
                value={category}
                onChange={handleSelect}
              >
                <option value="">Select Category</option>
                <option value="Breakfast">Breakfast</option>
                <option name="Lunch" value="Lunch">Lunch</option>
                <option value="Dinner">Dinner</option>
                <option value="Appetizer">Appetizer</option>
                <option value="Main">Main</option>
                <option value="Dessert">Dessert</option>
              </select>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <textarea
                  name="description"
                  className="materialize-textarea validate"
                  onChange={handleChange}
                  maxLength="255"
                  required
                />
                <label htmlFor="recipe_description">Description</label>
              </div>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <input
                  name="preparationTime"
                  type="number"
                  className="validate"
                  min="0"
                  onChange={handleChange}
                  required
                />
                <label htmlFor="recipe_prep_time">
                  Preparation Time (MINS)
                </label>
              </div>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <textarea
                  name="ingredients"
                  placeholder="Enter ingredients separated by comma"
                  className="materialize-textarea validate"
                  onChange={handleChange}
                  required
                />
                <label htmlFor="recipe_ingredients" className="active">
                  Ingredients
                </label>
              </div>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <textarea
                  name="directions"
                  placeholder="Enter directions separated by full stop"
                  className="materialize-textarea validate"
                  onChange={handleChange}
                  required
                />
                <label htmlFor="recipe_directions" className="active">
                  Directions
                </label>
              </div>
            </div>
            <div className="row center-align">
              <div className="col s12">
                <div className="dropzone-area">
                  <Dropzone
                    className="dropzone"
                    onDrop={handleDrop}
                    accept="image/*"
                    multiple={false}
                  >
                    {!imagePreview ?
                      <p className="center-align">
                        Drop your file or click here to upload
                        <i className="material-icons center large">camera</i>
                      </p> :
                      <img src={imagePreview} alt="" />}
                  </Dropzone>
                  <div className="row" />
                  {(imageUploading || isLoading) && <Spinner size="small" />}
                </div>
              </div>
            </div>
            {!isEmpty(error) &&
              <ErrorMessage message={error.message} />}
            <div className="row" />
            <div className="row center-align">
              <button
                id="add-recipe"
                name="add-recipe"
                type="submit"
                className={`btn btn-large waves-effect
                    waves-light indigo accent-2`}
                disabled={imageUploading || isLoading}
              >
                Add Recipe
              </button>
            </div>
          </form>
        </div>
      </Dialog>
    </div>
  );
};

AddRecipe.propTypes = propTypes;

export default AddRecipe;
