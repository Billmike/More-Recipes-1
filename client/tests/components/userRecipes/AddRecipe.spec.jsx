import React from 'react';

import AddRecipe from '../../../components/userRecipes/AddRecipe';

describe('AddRecipe component', () => {
  it('should render component', () => {
    const props = {
      category: 'Breakfast',
      open: true,
      handleClose: jest.fn(),
      handleChange: jest.fn(),
      handleSelect: jest.fn(),
      handleAddRecipe: jest.fn(),
      handleDrop: jest.fn(),
      imagePreview: '',
      userRecipes: {
        isLoading: false,
        imageUploading: false,
        error: {},
      },
    };
    const wrapper = shallow(<AddRecipe {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should show spinner when adding recipe', () => {
    const props = {
      category: 'Breakfast',
      open: true,
      handleClose: jest.fn(),
      handleChange: jest.fn(),
      handleSelect: jest.fn(),
      handleAddRecipe: jest.fn(),
      handleDrop: jest.fn(),
      imagePreview: 'Image file local url',
      userRecipes: {
        isLoading: true,
        imageUploading: true,
        error: {},
      },
    };
    const wrapper = shallow(<AddRecipe {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('Spinner').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should show error message if any', () => {
    const props = {
      category: 'Breakfast',
      open: true,
      handleClose: jest.fn(),
      handleChange: jest.fn(),
      handleSelect: jest.fn(),
      handleAddRecipe: jest.fn(),
      handleDrop: jest.fn(),
      imagePreview: 'Image file local url',
      userRecipes: {
        isLoading: false,
        imageUploading: false,
        error: {
          message: 'Error adding recipe',
        },
      },
    };
    const wrapper = shallow(<AddRecipe {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('Spinner').length).toEqual(0);
    expect(wrapper.find('ErrorMessage').length).toEqual(1);
    expect(wrapper.find('ErrorMessage').dive().text())
      .toEqual('Error adding recipe');
    expect(wrapper).toMatchSnapshot();
  });
});
