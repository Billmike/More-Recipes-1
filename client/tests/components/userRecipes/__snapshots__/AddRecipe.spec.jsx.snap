// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipe component should render component 1`] = `
<div>
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onClick={[MockFunction]}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={true}
        />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={true}
    modal={true}
    open={true}
    repositionOnUpdate={true}
    title="Add Recipe"
  >
    <div
      className="row"
      id="add-recipe"
    >
      <form
        className="col s12"
        onSubmit={[MockFunction]}
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              name="title"
              onChange={[MockFunction]}
              required={true}
              type="text"
            />
            <label
              htmlFor="recipe_name"
            >
              Title
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <select
            className="browser-default"
            onChange={[MockFunction]}
            value="Breakfast"
          >
            <option
              value=""
            >
              Select Category
            </option>
            <option
              value="Breakfast"
            >
              Breakfast
            </option>
            <option
              name="Lunch"
              value="Lunch"
            >
              Lunch
            </option>
            <option
              value="Dinner"
            >
              Dinner
            </option>
            <option
              value="Appetizer"
            >
              Appetizer
            </option>
            <option
              value="Main"
            >
              Main
            </option>
            <option
              value="Dessert"
            >
              Dessert
            </option>
          </select>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="description"
              onChange={[MockFunction]}
              required={true}
            />
            <label
              htmlFor="recipe_description"
            >
              Description
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              min="0"
              name="preparationTime"
              onChange={[MockFunction]}
              required={true}
              type="number"
            />
            <label
              htmlFor="recipe_prep_time"
            >
              Preparation Time (MINS)
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="ingredients"
              onChange={[MockFunction]}
              placeholder="Enter ingredients separated by comma"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_ingredients"
            >
              Ingredients
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="directions"
              onChange={[MockFunction]}
              placeholder="Enter directions separated by full stop"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_directions"
            >
              Directions
            </label>
          </div>
        </div>
        <div
          className="row center-align"
        >
          <div
            className="col s12"
          >
            <div
              className="dropzone-area"
            >
              <Dropzone
                accept="image/*"
                className="dropzone"
                disableClick={false}
                disablePreview={false}
                disabled={false}
                maxSize={Infinity}
                minSize={0}
                multiple={false}
                onDrop={[MockFunction]}
                preventDropOnDocument={true}
              >
                <p
                  className="center-align"
                >
                  Drop your file or click here to upload
                  <i
                    className="material-icons center large"
                  >
                    camera
                  </i>
                </p>
              </Dropzone>
              <div
                className="row"
              />
            </div>
          </div>
        </div>
        <div
          className="row"
        />
        <div
          className="row center-align"
        >
          <button
            className="btn btn-large waves-effect
                    waves-light indigo accent-2"
            disabled={false}
            id="add-recipe"
            name="add-recipe"
            type="submit"
          >
            Add Recipe
          </button>
        </div>
      </form>
    </div>
  </Dialog>
</div>
`;

exports[`AddRecipe component should show error message if any 1`] = `
<div>
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onClick={[MockFunction]}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={true}
        />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={true}
    modal={true}
    open={true}
    repositionOnUpdate={true}
    title="Add Recipe"
  >
    <div
      className="row"
      id="add-recipe"
    >
      <form
        className="col s12"
        onSubmit={[MockFunction]}
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              name="title"
              onChange={[MockFunction]}
              required={true}
              type="text"
            />
            <label
              htmlFor="recipe_name"
            >
              Title
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <select
            className="browser-default"
            onChange={[MockFunction]}
            value="Breakfast"
          >
            <option
              value=""
            >
              Select Category
            </option>
            <option
              value="Breakfast"
            >
              Breakfast
            </option>
            <option
              name="Lunch"
              value="Lunch"
            >
              Lunch
            </option>
            <option
              value="Dinner"
            >
              Dinner
            </option>
            <option
              value="Appetizer"
            >
              Appetizer
            </option>
            <option
              value="Main"
            >
              Main
            </option>
            <option
              value="Dessert"
            >
              Dessert
            </option>
          </select>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="description"
              onChange={[MockFunction]}
              required={true}
            />
            <label
              htmlFor="recipe_description"
            >
              Description
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              min="0"
              name="preparationTime"
              onChange={[MockFunction]}
              required={true}
              type="number"
            />
            <label
              htmlFor="recipe_prep_time"
            >
              Preparation Time (MINS)
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="ingredients"
              onChange={[MockFunction]}
              placeholder="Enter ingredients separated by comma"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_ingredients"
            >
              Ingredients
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="directions"
              onChange={[MockFunction]}
              placeholder="Enter directions separated by full stop"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_directions"
            >
              Directions
            </label>
          </div>
        </div>
        <div
          className="row center-align"
        >
          <div
            className="col s12"
          >
            <div
              className="dropzone-area"
            >
              <Dropzone
                accept="image/*"
                className="dropzone"
                disableClick={false}
                disablePreview={false}
                disabled={false}
                maxSize={Infinity}
                minSize={0}
                multiple={false}
                onDrop={[MockFunction]}
                preventDropOnDocument={true}
              >
                <img
                  alt=""
                  src="Image file local url"
                />
              </Dropzone>
              <div
                className="row"
              />
            </div>
          </div>
        </div>
        <ErrorMessage
          message="Error adding recipe"
        />
        <div
          className="row"
        />
        <div
          className="row center-align"
        >
          <button
            className="btn btn-large waves-effect
                    waves-light indigo accent-2"
            disabled={false}
            id="add-recipe"
            name="add-recipe"
            type="submit"
          >
            Add Recipe
          </button>
        </div>
      </form>
    </div>
  </Dialog>
</div>
`;

exports[`AddRecipe component should show spinner when adding recipe 1`] = `
<div>
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onClick={[MockFunction]}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={true}
        />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={true}
    modal={true}
    open={true}
    repositionOnUpdate={true}
    title="Add Recipe"
  >
    <div
      className="row"
      id="add-recipe"
    >
      <form
        className="col s12"
        onSubmit={[MockFunction]}
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              name="title"
              onChange={[MockFunction]}
              required={true}
              type="text"
            />
            <label
              htmlFor="recipe_name"
            >
              Title
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <select
            className="browser-default"
            onChange={[MockFunction]}
            value="Breakfast"
          >
            <option
              value=""
            >
              Select Category
            </option>
            <option
              value="Breakfast"
            >
              Breakfast
            </option>
            <option
              name="Lunch"
              value="Lunch"
            >
              Lunch
            </option>
            <option
              value="Dinner"
            >
              Dinner
            </option>
            <option
              value="Appetizer"
            >
              Appetizer
            </option>
            <option
              value="Main"
            >
              Main
            </option>
            <option
              value="Dessert"
            >
              Dessert
            </option>
          </select>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="description"
              onChange={[MockFunction]}
              required={true}
            />
            <label
              htmlFor="recipe_description"
            >
              Description
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              min="0"
              name="preparationTime"
              onChange={[MockFunction]}
              required={true}
              type="number"
            />
            <label
              htmlFor="recipe_prep_time"
            >
              Preparation Time (MINS)
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="ingredients"
              onChange={[MockFunction]}
              placeholder="Enter ingredients separated by comma"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_ingredients"
            >
              Ingredients
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea validate"
              name="directions"
              onChange={[MockFunction]}
              placeholder="Enter directions separated by full stop"
              required={true}
            />
            <label
              className="active"
              htmlFor="recipe_directions"
            >
              Directions
            </label>
          </div>
        </div>
        <div
          className="row center-align"
        >
          <div
            className="col s12"
          >
            <div
              className="dropzone-area"
            >
              <Dropzone
                accept="image/*"
                className="dropzone"
                disableClick={false}
                disablePreview={false}
                disabled={false}
                maxSize={Infinity}
                minSize={0}
                multiple={false}
                onDrop={[MockFunction]}
                preventDropOnDocument={true}
              >
                <img
                  alt=""
                  src="Image file local url"
                />
              </Dropzone>
              <div
                className="row"
              />
              <Spinner
                size="small"
              />
            </div>
          </div>
        </div>
        <div
          className="row"
        />
        <div
          className="row center-align"
        >
          <button
            className="btn btn-large waves-effect
                    waves-light indigo accent-2"
            disabled={true}
            id="add-recipe"
            name="add-recipe"
            type="submit"
          >
            Add Recipe
          </button>
        </div>
      </form>
    </div>
  </Dialog>
</div>
`;
